include/nbl/builtin/hlsl/emulated/emulated_float64_t_impl.hlsl

```diff
@@ -0,0 +1,327 @@
+    }
+};
+
+NBL_CONSTEXPR_INLINE_FUNC bool isZero(uint64_t val)
+{
+    return (val << 1) == 0;
+}
+
+inline uint64_t castToUint64WithFloat64BitPattern(uint64_t val)
+{
+    if (isZero(val))
+        return val;
+
+#ifndef __HLSL_VERSION
+    int exp = findMSB(val);
+#else
+    int exp = 63;
+    uint64_t mask = ieee754::traits<float64_t>::signMask;
+    while (!(val & mask))
+    {
+        --exp;
+        mask >>= 1;
+    }
+
+
+    //uint32_t2 valPacked = packUint64(val);
+    //int exp = valPacked.x ? firstbithigh(valPacked.x) + 32 : firstbithigh(valPacked.y);
+    //exp = 63 - exp;
+#endif
+    uint64_t mantissa;
+
+    int shiftCnt = 52 - exp;
+    if (shiftCnt >= 0)
+    {
+        mantissa = val << shiftCnt;
+    }
+    else
+    {
+        const int shiftCntAbs = -shiftCnt;
+        uint64_t roundingBit = 1ull << (shiftCnt - 1);
+        uint64_t stickyBitMask = roundingBit - 1;
+        uint64_t stickyBit = val & stickyBitMask;
+
+        mantissa = val >> shiftCntAbs;
+
+        if ((val & roundingBit) && (!stickyBit))
+        {
+            bool isEven = mantissa & 1;
+            if (!isEven)
+                mantissa++;
+        }
+        else if ((val & roundingBit) && (stickyBit || (mantissa & 1)))
+            val += roundingBit;
+
+        //val += (1ull << (shiftCnt)) - 1;
+        //mantissa = val >> shiftCntAbs;
+
+        if (mantissa & 1ull << 53)
+        {
+            mantissa >>= 1;
+            exp++;
+        }
+    }
+    mantissa &= ieee754::traits<float64_t>::mantissaMask;
+    const uint64_t biasedExp = uint64_t(ieee754::traits<float64_t>::exponentBias + exp) << ieee754::traits<float64_t>::mantissaBitCnt;
+
+    return biasedExp | mantissa;
+};
+
+inline uint64_t castToUint64WithFloat64BitPattern(int64_t val)
+{
+    const uint64_t sign = val & ieee754::traits<float64_t>::signMask;
+    const uint64_t absVal = uint64_t(abs(val));
+    return sign | castToUint64WithFloat64BitPattern(absVal);
+};
+
+NBL_CONSTEXPR_INLINE_FUNC uint32_t2 umulExtended(uint32_t lhs, uint32_t rhs)
+{
+    uint64_t product = uint64_t(lhs) * uint64_t(rhs);
+    uint32_t2 output;
+    output.x = uint32_t((product & 0xFFFFFFFF00000000) >> 32);
+    output.y = uint32_t(product & 0x00000000FFFFFFFFull);
+    return output;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t propagateFloat64NaN(uint64_t lhs, uint64_t rhs)
+{
+#if defined RELAXED_NAN_PROPAGATION
+    return lhs | rhs;
+#else
+
+    lhs |= 0x0008000000000000ull;
+    rhs |= 0x0008000000000000ull;
+    return glsl::mix(rhs, glsl::mix(lhs, rhs, tgmath::isnan(rhs)), tgmath::isnan(lhs));
+    return 0;
+#endif
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t flushDenormToZero(uint64_t extractedBiasedExponent, uint64_t value)
+{
+    return extractedBiasedExponent ? value : ieee754::extractSignPreserveBitPattern(value);
+}
+
+static inline int countLeadingZeros32(uint32_t val)
+{
+#ifndef __HLSL_VERSION
+    return 31 - findMSB(val);
+#else
+    return 31 - firstbithigh(val);
+#endif
+}
```

Comment:
should be somewhere in another header, named as `clz`

Reply:
removed